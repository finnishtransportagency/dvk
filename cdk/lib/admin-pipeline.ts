import * as cdk from 'aws-cdk-lib';
import { Arn, CfnOutput, SecretValue, Stack } from 'aws-cdk-lib';
import * as codepipeline from 'aws-cdk-lib/aws-codepipeline';
import * as codebuild from 'aws-cdk-lib/aws-codebuild';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import { ComputeType, LinuxBuildImage } from 'aws-cdk-lib/aws-codebuild';
import { Repository } from 'aws-cdk-lib/aws-ecr';
import { GitHubTrigger } from 'aws-cdk-lib/aws-codepipeline-actions';
import { Effect } from 'aws-cdk-lib/aws-iam';
interface AdminPipelineProps {
  env: string;
}
export class AdminPipeline extends Construct {
  constructor(scope: Stack, id: string, props: AdminPipelineProps) {
    super(scope, id);

    const account = cdk.Stack.of(this).account;

    const pipeline = new codepipeline.Pipeline(this, 'AdminPipeline', {
      crossAccountKeys: false,
    });
    const sourceOutput = new codepipeline.Artifact();
    const sourceAction = new cdk.aws_codepipeline_actions.GitHubSourceAction({
      actionName: 'GitHub_Source',
      owner: 'finnishtransportagency',
      repo: 'dvk',
      oauthToken: SecretValue.secretsManager('dev/dvk/github'),
      output: sourceOutput,
      branch: this.getBranch(props.env),
      trigger: GitHubTrigger.NONE,
    });

    pipeline.addStage({
      stageName: 'Source',
      actions: [sourceAction],
    });

    const adminBuildProject = new codebuild.PipelineProject(this, 'AdminBuild', {
      environment: {
        buildImage: LinuxBuildImage.fromEcrRepository(Repository.fromRepositoryName(this, 'DvkBuildImage', 'dvk-buildimage'), '1.0.3'),
      },
      buildSpec: codebuild.BuildSpec.fromObject({
        version: '0.2',
        phases: {
          install: {
            commands: ['echo Show node versions', 'node -v', 'npm -v'],
          },
          build: {
            commands: ['echo build admin app', 'cd admin', 'npm ci', 'BUILD_PATH=./build/yllapito PUBLIC_URL=/yllapito npm run build'],
          },
        },
        artifacts: {
          'base-directory': 'admin/build',
          files: '**/*',
        },
      }),
    });

    const buildOutput = new codepipeline.Artifact();

    pipeline.addStage({
      stageName: 'Build',
      actions: [
        new cdk.aws_codepipeline_actions.CodeBuildAction({
          actionName: 'BuildAdminApp',
          project: adminBuildProject,
          input: sourceOutput,
          outputs: [buildOutput],
        }),
      ],
    });

    const importedDistributionId = cdk.Fn.importValue('SquatDistribution' + props.env);

    // Create the build project that will invalidate the cache
    const invalidateBuildProject = new codebuild.PipelineProject(this, 'InvalidateProject', {
      environment: {
        buildImage: LinuxBuildImage.fromEcrRepository(Repository.fromRepositoryName(this, 'InvalidateBuildImage', 'dvk-buildimage'), '1.0.3'),
        computeType: ComputeType.SMALL,
      },
      buildSpec: codebuild.BuildSpec.fromObject({
        version: '0.2',
        phases: {
          build: {
            commands: [
              // eslint-disable-next-line no-template-curly-in-string
              'aws cloudfront create-invalidation --distribution-id ${CLOUDFRONT_ID} --paths "/*"',
            ],
          },
        },
      }),
      environmentVariables: {
        CLOUDFRONT_ID: { value: importedDistributionId },
      },
    });

    // Add Cloudfront invalidation permissions to the project
    const distributionArn = `arn:aws:cloudfront::${account}:distribution/${importedDistributionId}`;

    invalidateBuildProject.addToRolePolicy(
      new iam.PolicyStatement({
        resources: [distributionArn],
        actions: ['cloudfront:CreateInvalidation'],
      })
    );

    const importedBucketValue = cdk.Fn.importValue('AdminBucket' + props.env);

    // Create the build project that will delete files from s3 before deploy
    const emptyS3BuildProject = new codebuild.PipelineProject(this, 'EmptyS3Project', {
      environment: {
        buildImage: LinuxBuildImage.fromEcrRepository(Repository.fromRepositoryName(this, 'S3BuildImage', 'dvk-buildimage'), '1.0.3'),
        computeType: ComputeType.SMALL,
      },
      buildSpec: codebuild.BuildSpec.fromObject({
        version: '0.2',
        phases: {
          build: {
            commands: [
              // eslint-disable-next-line no-template-curly-in-string
              'aws s3 rm s3://${TARGET_BUCKET}/yllapito --recursive',
            ],
          },
        },
      }),
      environmentVariables: {
        TARGET_BUCKET: { value: importedBucketValue },
      },
    });

    const s3RootArn = Arn.format({
      region: '',
      service: 's3',
      resource: importedBucketValue,
      account: '',
      partition: 'aws',
    });

    const s3FolderArn = Arn.format({
      region: '',
      service: 's3',
      resource: importedBucketValue,
      resourceName: 'yllapito/*',
      account: '',
      partition: 'aws',
    });

    emptyS3BuildProject.addToRolePolicy(
      new iam.PolicyStatement({
        resources: [s3RootArn],
        actions: ['s3:ListBucket', 's3:GetBucketLocation'],
        effect: Effect.ALLOW,
      })
    );

    emptyS3BuildProject.addToRolePolicy(
      new iam.PolicyStatement({
        resources: [s3FolderArn],
        actions: ['s3:GetObject', 's3:GetObjectAcl', 's3:DeleteObject'],
        effect: Effect.ALLOW,
      })
    );

    pipeline.addStage({
      stageName: 'Deploy',
      actions: [
        new cdk.aws_codepipeline_actions.CodeBuildAction({
          actionName: 'DeleteFiles',
          project: emptyS3BuildProject,
          input: buildOutput,
          runOrder: 1,
        }),
        new cdk.aws_codepipeline_actions.S3DeployAction({
          actionName: 'S3Deploy',
          bucket: s3.Bucket.fromBucketName(this, 'Bucket', importedBucketValue.toString()),
          input: buildOutput,
          runOrder: 2,
        }),
        new cdk.aws_codepipeline_actions.CodeBuildAction({
          actionName: 'InvalidateCache',
          project: invalidateBuildProject,
          input: buildOutput,
          runOrder: 3,
        }),
      ],
    });

    new CfnOutput(this, 'PipelineName', {
      value: pipeline.pipelineName,
      description: 'Admin pipeline name',
      exportName: 'AdminPipeline-' + props.env,
    });

    new CfnOutput(this, 'PipelineARN', {
      value: pipeline.pipelineArn,
      description: 'Admin pipeline ARN',
      exportName: 'AdminPipeline-ARN-' + props.env,
    });
  }

  private getBranch(env: string): string {
    switch (env) {
      case 'prod':
        return 'prod';
      case 'test':
        return 'test';
      default:
        return 'main';
    }
  }
}
