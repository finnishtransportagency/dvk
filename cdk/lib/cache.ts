export type CloudFrontCacheKey =
  | 'aislocations'
  | 'aisvessels'
  | 'mareograph'
  | 'observation'
  | 'buoy'
  | 'marinewarning'
  | 'feature'
  | 'pilotroutes'
  | 'forecast';

export const FEATURE_CACHE_DURATION = 7200; // 2 hours

type CloudFrontCacheItem = {
  key: CloudFrontCacheKey;
  maxAge: number;
  staleWhileRevalidate: number;
  staleIfError: number;
};

const CloudFrontCacheItems: Array<CloudFrontCacheItem> = [
  {
    key: 'aislocations',
    maxAge: 2, // 2 seconds
    staleWhileRevalidate: 3, // 3 seconds
    staleIfError: 58, // 58 seconds (do not return location data more than 1 minute old)
  },
  {
    key: 'aisvessels',
    maxAge: 300, // 5 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 300, // 5 minutes (do not return vessel data more than 10 minutes old)
  },
  {
    key: 'mareograph',
    maxAge: 60, // 1 minute
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'observation',
    maxAge: 60, // 1 minute
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'buoy',
    maxAge: 60, // 1 minute
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'marinewarning',
    maxAge: 540, // 9 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'feature',
    maxAge: 7140, // 1 hour 59 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'pilotroutes',
    maxAge: 7140, // 1 hour 59 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'forecast',
    maxAge: 7140, // 1 hour 59 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
];

export function getAisCacheControlHeaders(key: 'aisvessels' | 'aislocations'): Record<string, string[]> {
  return getCloudFrontCacheControlHeaders(key);
}

export function isCloudFrontCacheKey(key: string): key is CloudFrontCacheKey {
  return CloudFrontCacheItems.some((item) => item.key === key );
}

export function getFeatureCacheControlHeaders(): Record<string, string[]> {
  return getCloudFrontCacheControlHeaders('feature');
}

export function getCloudFrontCacheControlHeaders(key: CloudFrontCacheKey): Record<string, string[]> {
  const cacheItem = CloudFrontCacheItems.find((item) => item.key === key );
  if (!cacheItem) {
    return getFeatureCacheControlHeaders();
  }
  const maxAge = cacheItem.maxAge;
  const staleWhileRevalidate = cacheItem.staleWhileRevalidate;
  const staleIfError = cacheItem.staleIfError;

  return {
    'Cache-Control': ['max-age=' + maxAge + ', ' + 'stale-while-revalidate=' + staleWhileRevalidate + ', ' + 'stale-if-error=' + staleIfError],
  };
}
