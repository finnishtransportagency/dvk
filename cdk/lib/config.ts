import { Construct } from 'constructs';
import * as ssm from 'aws-cdk-lib/aws-ssm';

class Config {
  private scope: Construct;

  constructor(scope: Construct) {
    this.scope = scope;
  }

  static isPermanentEnvironment() {
    return ['dev', 'test', 'prod'].indexOf(Config.getEnvironment()) >= 0;
  }

  static getEnvironment(): string {
    if (process.env.ENVIRONMENT) {
      return process.env.ENVIRONMENT;
    }
    throw new Error(Config.errorMessage('ENVIRONMENT'));
  }

  private static errorMessage(variable: string): string {
    return `Environment variable ${variable} missing, run '. ${__dirname}/../bin/setenv.sh' to set it`;
  }

  static getPublicIP(): string {
    if (process.env.PUBLIC_IP) {
      return process.env.PUBLIC_IP;
    }
    throw new Error(Config.errorMessage('PUBLIC_IP'));
  }

  public getStringParameter(parameterName: string): string {
    return ssm.StringParameter.valueForStringParameter(this.scope, `/${Config.getEnvironment()}/${parameterName}`);
  }

  public getGlobalStringParameter(parameterName: string): string {
    return ssm.StringParameter.valueForStringParameter(this.scope, `/${parameterName}`);
  }
}

export default Config;
