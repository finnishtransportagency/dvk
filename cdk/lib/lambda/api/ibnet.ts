import { Feature, FeatureCollection, GeoJsonProperties, Geometry } from 'geojson';
import { Dirway, DirwayPoint, IBNetApiResponse } from './apiModels';
import { fetchIBNetApi } from './axios';
import { log } from '../logger';
import { Position } from '@turf/helpers';

export const DIRWAY_PATH = 'dirway';
export const DIRWAY_POINT_PATH = 'dirwaypoint';

function addDirwayFeatures(features: Feature<Geometry, GeoJsonProperties>[], dirwayPoints: DirwayPoint[], dirways: Dirway[]) {
  const dirwayMap = new Map<string, DirwayPoint[]>();
  for (const p of dirwayPoints) {
    if (!p.deleted) {
      dirwayMap.set(p.dirway_id, (dirwayMap.get(p.dirway_id) ?? []).concat(p));
    }
  }
  for (const points of dirwayMap.values()) {
    const id = points[0].dirway_id;
    const coordinates = points.sort((a, b) => a.order_num - b.order_num).map((p) => [p.longitude, p.latitude] as Position);
    const dirway = dirways.filter((d) => !d.deleted).find((d) => d.id === id);

    const properties = {
      id: id,
      name: dirway?.name,
      description: dirway?.description,
      updated: dirway?.change_time,
      featureType: 'dirway',
    };

    features.push({
      type: 'Feature',
      id: id,
      geometry: { type: 'LineString', coordinates: coordinates },
      properties: properties,
    });
  }
}

export async function fetchDirways(): Promise<FeatureCollection> {
  const features: Feature<Geometry, GeoJsonProperties>[] = [];

  // First fetch root of the API to get max toRV value
  const apiResponse: IBNetApiResponse | null = await fetchIBNetApi<IBNetApiResponse>();
  const { toRv } = apiResponse ?? {};

  if (toRv) {
    const params = { from: '0', to: toRv };
    const dirways: Dirway[] | null = await fetchIBNetApi<Dirway[]>(DIRWAY_PATH, params);
    const dirwayPoints: DirwayPoint[] | null = await fetchIBNetApi<DirwayPoint[]>(DIRWAY_POINT_PATH, params);

    if (dirways && dirwayPoints) {
      addDirwayFeatures(features, dirwayPoints, dirways);
    }
  } else {
    log.error('Could not get toRv from IBNetApi, not fetching dirway data');
  }

  const collection: FeatureCollection = {
    type: 'FeatureCollection',
    features: features,
  };
  return collection;
}
