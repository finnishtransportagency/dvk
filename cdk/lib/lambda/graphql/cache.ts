import { DeleteObjectsCommand, GetObjectCommand, ObjectIdentifier, PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { getEnvironment } from '../environment';
import { log } from '../logger';
import { Readable } from 'stream';

const s3Client = new S3Client({ region: 'eu-west-1' });

const AIS_LOCATION_CACHE = {
  MAX_AGE: 2, // 2 seconds
  STALE_WHILE_REVALIDATE: 3, // 3 seconds
  STALE_IF_ERROR: 58, // 58 seconds (do not return location data more than 1 minute old)
};
const AIS_VESSEL_CACHE = {
  MAX_AGE: 300, // 5 minutes
  STALE_WHILE_REVALIDATE: 60, // 1 minute
  STALE_IF_ERROR: 300, // 5 minutes (do not return vessel data more than 10 minutes old)
};

export const FEATURE_CACHE_DURATION = 7200; // 2 hours

function getCacheBucketName() {
  return `featurecache-${getEnvironment()}`;
}

export function getFeatureCacheDuration(key: string) {
  if (key === 'aislocations') {
    return AIS_LOCATION_CACHE.MAX_AGE;
  } else if (key === 'aisvessels') {
    return AIS_VESSEL_CACHE.MAX_AGE;
  } else {
    return FEATURE_CACHE_DURATION;
  }
}

export function getCacheControlHeaders(key: string): Record<string, string[]> {
  if (key === 'aislocations') {
    return {
      'Cache-Control': [
        'max-age=' +
          AIS_LOCATION_CACHE.MAX_AGE +
          ', ' +
          'stale-while-revalidate=' +
          AIS_LOCATION_CACHE.STALE_WHILE_REVALIDATE +
          ', ' +
          'stale-if-error=' +
          AIS_LOCATION_CACHE.STALE_IF_ERROR,
      ],
    };
  } else if (key === 'aisvessels') {
    return {
      'Cache-Control': [
        'max-age=' +
          AIS_VESSEL_CACHE.MAX_AGE +
          ', ' +
          'stale-while-revalidate=' +
          AIS_VESSEL_CACHE.STALE_WHILE_REVALIDATE +
          ', ' +
          'stale-if-error=' +
          AIS_VESSEL_CACHE.STALE_IF_ERROR,
      ],
    };
  } else {
    return {};
  }
}

export async function cacheResponse(key: string, response: object | string) {
  const cacheDurationSeconds = getFeatureCacheDuration(key);
  const expires = new Date();
  expires.setTime(expires.getTime() + cacheDurationSeconds * 1000);
  const command = new PutObjectCommand({
    Key: key,
    Bucket: getCacheBucketName(),
    Expires: expires,
    Body: typeof response === 'string' ? response : JSON.stringify(response),
  });
  await s3Client.send(command);
}

export type CacheResponse = {
  expired: boolean;
  data?: string;
  lastModified?: string;
};

export async function getFromCache(key: string): Promise<CacheResponse> {
  try {
    const data = await s3Client.send(
      new GetObjectCommand({
        Key: key,
        Bucket: getCacheBucketName(),
      })
    );
    if (data.Body) {
      log.debug(`returning ${key} from cache`);
      const streamToString = (stream: Readable): Promise<string> =>
        new Promise((resolve, reject) => {
          const chunks: Uint8Array[] = [];
          stream.on('data', (chunk: Uint8Array) => chunks.push(chunk));
          stream.on('error', reject);
          stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
        });
      const expires = data.ExpiresString ? Date.parse(data.ExpiresString) : undefined;
      return {
        expired: expires !== undefined && !isNaN(expires) && expires < Date.now(),
        data: await streamToString(data.Body as Readable),
        lastModified: data.LastModified?.toUTCString(),
      };
    }
  } catch (e) {
    // errors ignored also not found
  }
  return { expired: true };
}

export async function deleteCacheObjects(keys: string[]) {
  const objects: ObjectIdentifier[] = keys.map((key) => {
    return { Key: key };
  });
  const response = await s3Client.send(
    new DeleteObjectsCommand({
      Bucket: getCacheBucketName(),
      Delete: {
        Objects: objects,
      },
    })
  );
  log.debug(`Deleted cache objects ${keys}: ${response.Deleted?.length}/${keys.length}`);
  if (response.Errors?.length) {
    log.error(response.Errors);
  }
}
