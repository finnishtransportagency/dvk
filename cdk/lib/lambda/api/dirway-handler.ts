import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import { toBase64Response } from '../util';
import { getHeaders } from '../environment';
import { fetchDirways } from './ibnet';
import { getFeatureCacheControlHeaders } from '../../cache';

export const DIRWAYS_KEY = 'dirways';

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `dirways()`);
  let base64Response: string | undefined;
  let statusCode = 200;
  try {
    const dirwayData = await fetchDirways();
    log.debug('dirway data: %d', dirwayData.features.length);
    base64Response = await toBase64Response(dirwayData);
  } catch (e) {
    base64Response = undefined;
    statusCode = 503;
  }

  return {
    statusCode,
    body: base64Response,
    isBase64Encoded: true,
    multiValueHeaders: {
      ...getHeaders(),
      ...getFeatureCacheControlHeaders(DIRWAYS_KEY),
      'Content-Type': ['application/geo+json'],
    },
  };
};
