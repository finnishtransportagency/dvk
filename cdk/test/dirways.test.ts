import { mockALBEvent } from './mocks';
import { gunzip } from 'zlib';
import assert from 'assert';
import { FeatureCollection } from 'geojson';
import { DIRWAY_PATH, DIRWAY_POINT_PATH } from '../lib/lambda/api/ibnet';
import { handler } from '../lib/lambda/api/dirway-handler';
import { getFeatureCacheControlHeaders } from '../lib/cache';

const DIRWAYS_KEY = 'dirways';

jest.mock('../lib/lambda/environment', () => ({
  getEnvironment: () => 'mock',
  isPermanentEnvironment: () => false,
  getHeaders: () => {},
}));

const apiRootResponse = {
  rvEndpoints: ['icebreaker_status', 'dirwaypoint', 'location', 'dirway', 'restriction', 'source', 'aistarget', 'queue', 'activity', 'vessel'],
  toRv: 9420308526,
};

const dirwayResponse = [
  {
    rv: 7962700619,
    change_time: '2024-01-17T11:59:39.710Z',
    id: 'dirway-5_36000001',
    name: 'IBNEXT Center 0117 13:59:35',
    description: '',
  },
  {
    rv: 8845046484,
    change_time: '2024-04-18T09:25:31.446Z',
    id: 'dirway-771_36000005',
    name: 'Finnish icebreaker 1 0418 12:24:44',
    description: 'tt',
  },
  {
    rv: 7978396437,
    change_time: '2024-01-19T07:15:36.424Z',
    deleted: true,
    id: 'dirway-771_36000001',
  },
];

const dirwayPointResponse = [
  {
    rv: 7962700535,
    change_time: '2024-01-17T11:59:39.493Z',
    id: 'dirwaypoint-5_36000004',
    dirway_id: 'dirway-5_36000001',
    order_num: 2,
    name: 'wp3',
    latitude: 64.119,
    longitude: 21.5002,
  },
  {
    rv: 8845046481,
    change_time: '2024-04-18T09:25:31.241Z',
    id: 'dirwaypoint-771_36000024',
    dirway_id: 'dirway-771_36000005',
    order_num: 6,
    name: 'wp7',
    latitude: 65.1338,
    longitude: 22.266,
  },
  {
    rv: 7962700542,
    change_time: '2024-01-17T11:59:39.528Z',
    id: 'dirwaypoint-5_36000005',
    dirway_id: 'dirway-5_36000001',
    order_num: 3,
    name: 'wp4',
    latitude: 64.1813,
    longitude: 21.3712,
  },
  {
    rv: 7978396443,
    change_time: '2024-01-19T07:15:36.447Z',
    deleted: true,
    id: 'dirwaypoint-771_36000005',
    dirway_id: 'dirway-771_36000001',
  },
  {
    rv: 8845046463,
    change_time: '2024-04-18T09:25:31.196Z',
    id: 'dirwaypoint-771_36000018',
    dirway_id: 'dirway-771_36000005',
    order_num: 0,
    name: 'Nordvalen',
    latitude: 63.5254,
    longitude: 20.8039,
  },
  {
    rv: 8845046478,
    change_time: '2024-04-18T09:25:31.241Z',
    id: 'dirwaypoint-771_36000023',
    dirway_id: 'dirway-771_36000005',
    order_num: 5,
    name: 'wp6',
    latitude: 64.7053,
    longitude: 21.9253,
  },
  {
    rv: 8845046466,
    change_time: '2024-04-18T09:25:31.231Z',
    id: 'dirwaypoint-771_36000019',
    dirway_id: 'dirway-771_36000005',
    order_num: 1,
    name: 'wp2',
    latitude: 63.5645,
    longitude: 21.3041,
  },
  {
    rv: 7962700513,
    change_time: '2024-01-17T11:59:39.353Z',
    id: 'dirwaypoint-5_36000002',
    dirway_id: 'dirway-5_36000001',
    order_num: 0,
    name: 'wp1',
    latitude: 64.1058,
    longitude: 21.308,
  },
  {
    rv: 7978396440,
    change_time: '2024-01-19T07:15:36.447Z',
    deleted: true,
    id: 'dirwaypoint-771_36000006',
    dirway_id: 'dirway-771_36000001',
  },
  {
    rv: 8845046472,
    change_time: '2024-04-18T09:25:31.236Z',
    id: 'dirwaypoint-771_36000021',
    dirway_id: 'dirway-771_36000005',
    order_num: 3,
    name: 'wp4',
    latitude: 64.1216,
    longitude: 21.7054,
  },
  {
    rv: 7978396446,
    change_time: '2024-01-19T07:15:36.447Z',
    deleted: true,
    id: 'dirwaypoint-771_36000004',
    dirway_id: 'dirway-771_36000001',
  },
  {
    rv: 8845046475,
    change_time: '2024-04-18T09:25:31.236Z',
    id: 'dirwaypoint-771_36000022',
    dirway_id: 'dirway-771_36000005',
    order_num: 4,
    name: 'wp5',
    latitude: 64.4529,
    longitude: 21.9747,
  },
  {
    rv: 8845046469,
    change_time: '2024-04-18T09:25:31.231Z',
    id: 'dirwaypoint-771_36000020',
    dirway_id: 'dirway-771_36000005',
    order_num: 2,
    name: 'wp3',
    latitude: 63.7474,
    longitude: 21.7494,
  },
  {
    rv: 7962700531,
    change_time: '2024-01-17T11:59:39.483Z',
    id: 'dirwaypoint-5_36000003',
    dirway_id: 'dirway-5_36000001',
    order_num: 1,
    name: 'wp2',
    latitude: 64.0758,
    longitude: 21.3959,
  },
];

async function parseResponse(body: string): Promise<FeatureCollection> {
  const response = new Promise<Error | Buffer>((resolve, reject) =>
    gunzip(Buffer.from(body, 'base64'), (err, data) => {
      if (err) {
        reject(err);
      }
      resolve(data);
    })
  );
  return JSON.parse((await response).toString()) as FeatureCollection;
}

let throwError = false;
jest.mock('../lib/lambda/api/axios', () => ({
  fetchIBNetApi: (path?: string) => {
    if (throwError) {
      throw new Error('Fetching from IBNet api failed');
    }
    if (!path) {
      return apiRootResponse;
    } else if (path === DIRWAY_PATH) {
      return dirwayResponse;
    } else if (path === DIRWAY_POINT_PATH) {
      return dirwayPointResponse;
    }
    return null;
  },
}));

beforeEach(() => {
  jest.resetAllMocks();
  throwError = false;
});

it('should get dirways from api', async () => {
  const response = await handler(mockALBEvent(DIRWAYS_KEY));
  assert(response.body);
  const responseObj = await parseResponse(response.body);
  expect(responseObj.features.length).toBe(2);
  expect(responseObj).toMatchSnapshot();
});

it('should get internal server error when api call fails', async () => {
  throwError = true;
  const response = await handler(mockALBEvent(DIRWAYS_KEY));
  expect(response.statusCode).toBe(503);
});

it('should return right cache headers', async () => {
  const response = await handler(mockALBEvent(DIRWAYS_KEY));
  assert(response.body);
  const headers = getFeatureCacheControlHeaders(DIRWAYS_KEY)?.['Cache-Control'];
  expect(response?.multiValueHeaders?.['Cache-Control']).toStrictEqual(headers);
});
