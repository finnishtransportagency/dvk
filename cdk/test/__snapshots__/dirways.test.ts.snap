// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get dirways from api 1`] = `
{
  "features": [
    {
      "geometry": {
        "coordinates": [
          [
            20.8039,
            63.5254,
          ],
          [
            21.3041,
            63.5645,
          ],
          [
            21.7494,
            63.7474,
          ],
          [
            21.7054,
            64.1216,
          ],
          [
            21.9747,
            64.4529,
          ],
          [
            21.9253,
            64.7053,
          ],
          [
            22.266,
            65.1338,
          ],
        ],
        "type": "LineString",
      },
      "id": "dirway-771_36000005",
      "properties": {
        "description": "tt",
        "featureType": "dirway",
        "id": "dirway-771_36000005",
        "name": "Finnish icebreaker 1 0418 12:24:44",
        "points": [
          {
            "coordinates": [
              20.8039,
              63.5254,
            ],
            "orderNumber": 0,
          },
          {
            "coordinates": [
              21.3041,
              63.5645,
            ],
            "orderNumber": 1,
          },
          {
            "coordinates": [
              21.7494,
              63.7474,
            ],
            "orderNumber": 2,
          },
          {
            "coordinates": [
              21.7054,
              64.1216,
            ],
            "orderNumber": 3,
          },
          {
            "coordinates": [
              21.9747,
              64.4529,
            ],
            "orderNumber": 4,
          },
          {
            "coordinates": [
              21.9253,
              64.7053,
            ],
            "orderNumber": 5,
          },
          {
            "coordinates": [
              22.266,
              65.1338,
            ],
            "orderNumber": 6,
          },
        ],
        "updated": "2024-04-18T09:25:31.446Z",
      },
      "type": "Feature",
    },
    {
      "geometry": {
        "coordinates": [
          [
            21.308,
            64.1058,
          ],
          [
            21.3959,
            64.0758,
          ],
          [
            21.5002,
            64.119,
          ],
          [
            21.3712,
            64.1813,
          ],
        ],
        "type": "LineString",
      },
      "id": "dirway-5_36000001",
      "properties": {
        "description": "",
        "featureType": "dirway",
        "id": "dirway-5_36000001",
        "name": "IBNEXT Center 0117 13:59:35",
        "points": [
          {
            "coordinates": [
              21.308,
              64.1058,
            ],
            "orderNumber": 0,
          },
          {
            "coordinates": [
              21.3959,
              64.0758,
            ],
            "orderNumber": 1,
          },
          {
            "coordinates": [
              21.5002,
              64.119,
            ],
            "orderNumber": 2,
          },
          {
            "coordinates": [
              21.3712,
              64.1813,
            ],
            "orderNumber": 3,
          },
        ],
        "updated": "2024-01-17T11:59:39.710Z",
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;
